src/java/org/apache/cassandra/locator/ReplicaLayout.java

```diff
@@ -0,0 +1,390 @@
+ * Encapsulates knowledge about the ring necessary for performing a specific operation, with static accessors
+ * for building the relevant layout.
+ *
+ * Constitutes:
+ *  - the 'natural' replicas replicating the range or token relevant for the operation
+ *  - if for performing a write, any 'pending' replicas that are taking ownership of the range, and must receive updates
+ *  - the 'selected' replicas, those that should be targeted for any operation
+ *  - 'all' replicas represents natural+pending
+ *
+ * @param <E> the type of Endpoints this ReplayLayout holds (either EndpointsForToken or EndpointsForRange)
+ * @param <L> the type of itself, including its type parameters, for return type of modifying methods
+ */
+public abstract class ReplicaLayout<E extends Endpoints<E>, L extends ReplicaLayout<E, L>>
+{
+    private volatile E all;
+    protected final E natural;
+    protected final E pending;
+    protected final E selected;
+
+    protected final Keyspace keyspace;
+    protected final ConsistencyLevel consistencyLevel;
+
+    private ReplicaLayout(Keyspace keyspace, ConsistencyLevel consistencyLevel, E natural, E pending, E selected)
+    {
+        this(keyspace, consistencyLevel, natural, pending, selected, null);
+    }
+
+    private ReplicaLayout(Keyspace keyspace, ConsistencyLevel consistencyLevel, E natural, E pending, E selected, E all)
+    {
+        assert selected != null;
+        assert pending == null || !Endpoints.haveConflicts(natural, pending);
+        this.keyspace = keyspace;
+        this.consistencyLevel = consistencyLevel;
+        this.natural = natural;
+        this.pending = pending;
+        this.selected = selected;
+        // if we logically have no pending endpoints (they are null), then 'all' our endpoints are natural
+        if (all == null && pending == null)
+            all = natural;
+        this.all = all;
+    }
+
+    public Replica getReplicaFor(InetAddressAndPort endpoint)
+    {
+        return natural.byEndpoint().get(endpoint);
+    }
+
+    public E natural()
+    {
+        return natural;
+    }
+
+    public E all()
+    {
+        E result = all;
+        if (result == null)
+            all = result = Endpoints.concat(natural, pending);
+        return result;
+    }
+
+    public E selected()
+    {
+        return selected;
+    }
+
+    /**
+     * @return the pending replicas - will be null for read layouts
+     * TODO: ideally we would enforce at compile time that read layouts have no pending to access
+     */
+    public E pending()
+    {
+        return pending;
+    }
+
+    public Keyspace keyspace()
+    {
+        return keyspace;
+    }
+
+    public ConsistencyLevel consistencyLevel()
+    {
+        return consistencyLevel;
+    }
+
+    abstract public L withSelected(E replicas);
+
+    abstract public L withConsistencyLevel(ConsistencyLevel cl);
+
+    public L forResponded(Iterable<InetAddressAndPort> endpoints)
+    {
+        // Preserve insertion order
```

Comment:
Having a little trouble following where insertion order is preserved, also why the extra copy for select.keep? Can't it take the iterable?

Reply:
OK, but it allocates a copy and I don't think it needs to you can pass the iterable to keep?

Reply:
The keep implementation that accepts an Iterable re-orders the Replica into the same order as the Iterable.  Perhaps we should rename it to select(), or just update the java doc, because it is counterintuitive.

Reply:
Ah I get it now.

Reply:
I think we should rename one of the two methods, thinking about it, because there's another place we invoke keep on the Iterable and *require* that it reorders the to the Iterable.  So it's super counterintuitive.

We could rename it to select() for consistency with the Select that works by adding predicates